module load varsim/0.5.1
module load jdk/1.8.0_25
module load R
module list

HOME="/gcc/groups/gcc/tmp01/rkanninga/vcf_compare/"
TRUTH="NextGene"
COMPARE="3.1.2"
OUTPUT="/gcc/groups/gcc/tmp01/rkanninga/vcf_compare/output"

if [ ! -d $OUTPUT ] 
then
	mkdir $OUTPUT
fi

THISDIR=`pwd`


for SAMPLE in 20060713_160511_DNA084315_407824_NGS1544 20060713_160538_DNA084314_407825_NGS1544 20060713_3950330_DNA084316_407822_NGS1544 20101604_567897_DNA083007_401873_NGS1544 20122050_191492_DNA084726_409419_NGS1544 20122050_4849448_DNA084730_409416_NGS1544 20122050_6033865_DNA084731_409412_NGS1544 20123629_3740522_DNA081585_409785_NGS1544 20123629_9321210_DNA084783_409779_NGS1544 20123629_9981455_DNA081584_409782_NGS1544 20130137_8006244_DNA062097_409571_NGS1544 20130137_8006244_DNA082392_409564_NGS1544 20130137_8646967_DNA062098_409566_NGS1544 20146366_9837713_DNA080106_409027_NGS1544 20146366_9975089_DNA081487_411063_NGS1544 20146366_9975097_DNA081488_411061_NGS1544 20150785_7498935_DNA082165_405956_NGS1544 20151485_38409_DNA085146_410999_NGS1544 20151485_9405717_DNA084718_409397_NGS1544 20151485_9959434_DNA084717_409396_NGS1544 20151569_218823_DNA085176_411080_NGS1544 20151569_218831_DNA085180_411083_NGS1544 20151569_9992236_DNA084463_410385_NGS1544 20152036_193516_DNA084786_409604_NGS1544 20152036_5804325_DNA084787_409602_NGS1544 20152036_9688769_DNA082942_409793_NGS1544 20152093_211540_DNA085084_411307_NGS1544 20152093_3116092_DNA082551_411305_NGS1544 20152093_6217459_DNA082550_411306_NGS1544 20152314_2972167_DNA083683_411105_NGS1544 20152314_810305_DNA085166_411076_NGS1544 20152314_810321_DNA085167_411079_NGS1544 
do
	echo "S1:${TRUTH}/intersect/${SAMPLE}*.intersect.vcf"  
	echo "S2:${COMPARE}/${SAMPLE}*.intersect.vcf"
	mkdir -p ${OUTPUT}/${SAMPLE}/

	cd /gcc/tools/varsim_0.5.1
	java -jar VarSim.jar vcfcompare -true_vcf ${HOME}/${TRUTH}/intersect/${SAMPLE}*intersect.vcf \
	-prefix ${OUTPUT}/${SAMPLE}/${SAMPLE}_${TRUTH}_VS_${COMPARE} \
	${HOME}/${COMPARE}/intersect/${SAMPLE}*intersect.vcf > ${OUTPUT}/${SAMPLE}/${SAMPLE}_${TRUTH}_VS_${COMPARE}.stats 2>&1
	
	echo "${OUTPUT}/${SAMPLE}/${SAMPLE}_${TRUTH}_VS_${COMPARE}.stats"
	perl -pi -e 's|Num new variants read.*|Num new variants read \nALLVARIANTS|' ${OUTPUT}/${SAMPLE}/${SAMPLE}_${TRUTH}_VS_${COMPARE}.stats 	
	
	Rscript /gcc/groups/gcc/home/gvdvries/Varsim_compare_test/make_plots_v2.R 'ALLVARIANTS' ${OUTPUT}/${SAMPLE}/${SAMPLE}_${TRUTH}_VS_${COMPARE}.stats "${TRUTH} vs ${COMPARE}"
	Rscript /gcc/groups/gcc/home/gvdvries/Varsim_compare_test/make_plots_v2.R 'SNP' ${OUTPUT}/${SAMPLE}/${SAMPLE}_${TRUTH}_VS_${COMPARE}.stats "${TRUTH} vs ${COMPARE}"
	Rscript /gcc/groups/gcc/home/gvdvries/Varsim_compare_test/make_plots_v2.R 'Insertion' ${OUTPUT}/${SAMPLE}/${SAMPLE}_${TRUTH}_VS_${COMPARE}.stats "${TRUTH} vs ${COMPARE}"
	Rscript /gcc/groups/gcc/home/gvdvries/Varsim_compare_test/make_plots_v2.R 'Deletion' ${OUTPUT}/${SAMPLE}/${SAMPLE}_${TRUTH}_VS_${COMPARE}.stats "${TRUTH} vs ${COMPARE}"

done

